### PR TITLE
[Stan 2.33] Make expiring deprecations into errors

### PR DESCRIPTION
**Pull request by WardBrian**
This adds some (primarily temporary) code which is modeled off the current `Deprecation_analysis` model but is used to generate errors rather than warnings. 

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [x] If a user-facing facing change was made, the documentation PR is here: https://github.com/stan-dev/docs/pull/625
    
## Release notes

The following deprecations have been turned into errors this version:

- The old array syntax, e.g. `int arr[5];`. Use `array[5] int arr;` instead.
- Distribution functions ending in `_log`. Use either `_lpdf` or `_lpmf`.
- The functions `binomial_coefficient_log`, `multiply_log`, and `cov_exp_quad`. Use `lchoose`, `lmultiply`, and `gp_exp_quad_cov` respectively.
- The `if_else` function. Use the ternary operator  `cond ? true_value : false_value`
- Use of CDFs with a `,` between the first and second argument. Use a `|`. 
- Comments beginning with `#`. Use `//`.
- Use of `<-` for assignment. Use `=`.
- The `increment_log_prob` function. Use the `target +=` statement.
- The `get_lp()` function. Use `target()`.
- The use of nested multi-indices on the left hand side of an assignment statement.

For this version, these can all be automatically updated with the `--canonicalize=deprecations` argument to the autoformatter. This is not guaranteed to work for versions _following_ this one.

Additionally, the following identifiers are now reserved words: `array`, `offset`, `multiplier`, `lower`, and `upper`.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/stanc/stanc.ml
Comment (nhuurre) [inline src/stanc/stanc.ml:224]: Do we document `stanc3` exit codes? Do we use other codes from `sysexits.h`?

I don't know what `sysexits.h` is but googling a bit I found https://man.netbsd.org/sysexits.3 and it looks to me that `EX_DATAERR` may be more appropriate than `EX_UNAVAILABLE` here

|code|description|
|-------|----------|
| EX_DATAERR | The input data was incorrect in some way.  This should only be used for user's data and not system files. |
| EX_UNAVAILABLE | A service is unavailable.  This can occur if a support program or file does not exist.  This can also be used as a catchall message when something you wanted to do does not work, but you do not know why. |

File: src/stancjs/stancjs.ml
Comment (nhuurre) [inline src/stancjs/stancjs.ml:61]: Was the previous absence of `deprecation_warnings` here an oversight?
And why don't you handle removals? Is `stancjs` expected to just allow deprecated features?
Feels very wrong that the only change to stancjs tests is to dismiss a pound-comment warning.

**Review by nhuurre — COMMENTED** [review]
You should change `Middle.Fun_kind.suffix_from_name` too since it makes use of the deprecated `_log` suffix.

The new `Deprecation_removals.ml` is mostly a copy of `Deprecation_analysis.ml`, right? What's your opinion on keeping them together, like, instead of `Deprecation_analysis.collect_warnings` you'd have `Deprecation_analysis.collect_warnings_and_removals` that returns two lists?

File: src/frontend/Typechecker.ml
Comment (WardBrian) [reply src/frontend/Typechecker.ml:1080]: You're right, this could be much simpler. I think the reason we needed to search before is that there could potentially be a `_log` and an `_lpdf`, one of which was correct and one of which wasn't? Or perhaps it was always over-complicated.

The return type is always going to be correct because of the `verify_fundef_dist_rt` check which occurs earlier

File: src/frontend/Input_warnings.ml
Comment (WardBrian) [reply src/frontend/Input_warnings.ml:47]: I would rather re-add it if/when it is needed than leave it in the source as dead code personally

File: src/stancjs/stancjs.ml
Comment (WardBrian) [reply src/stancjs/stancjs.ml:61]: Previous absence does seem to be an oversight.

Agreed, this was something I forgot to come back to. Stanc3js currently assumes any errors that occur are specifically of type `Errors.t`, which I didn't want to edit for this purpose, so it will need some tweaking

File: src/stanc/stanc.ml
Comment (WardBrian) [reply src/stanc/stanc.ml:224]: Currently the only exit codes we use are 0, 1, and 127 (only used if you don't pass a model on the command line). I figured it could be nice to pass a specific exit code for this failure since then wrappers like cmdstanr/py have the option of providing very specific messages. I'm happy for it to be any one of the numeric codes, though.

File: src/middle/Stan_math_signatures.ml
Comment (WardBrian) [reply src/middle/Stan_math_signatures.ml:255]: I agree it's unfortunate. I'd rather avoid the special case and have it consistently be true that you can only use `_lpdf` or `_lpmf`, especially since it is deprecated until "3.0", which could be a very long time from now.

**Comment by nhuurre** [comment]
> I like having it separate because I'm optimizing this code for deletion ... the idea is that most of this module will only exist for one version.

In that case it makes sense to keep them separate but I'd consider re-integrating them in the next version.

File: src/middle/Stan_math_signatures.ml
Comment (WardBrian) [reply src/middle/Stan_math_signatures.ml:255]: The idea is that they are "too breaking". We can't (or perhaps never tried to) provide an automated way of translating existing code to the new recommendation, so we don't want to remove them without a full major version bump.

**Review by nhuurre — APPROVED** [review]

**Comment by WardBrian** [comment]
Thanks @nhuurre. I'm going to wait to merge until I write the documentation PR early next week

**Comment by WardBrian** [comment]
@nhuurre while writing up the doc I realized the nested lvalue issue is also expiring. Care to take one last glance?

**Review by nhuurre — APPROVED** [review]
Looks good.

**Comment by WardBrian** [comment]
Updating now following #1303

We will not merge this until during the 2.33 development cycle

**Comment by WardBrian** [comment]
@nhuurre mind taking another look at this after the various merges that have been necessary?

**Review by nhuurre — APPROVED** [review]
Still looks good.

**Pull request merged**
